### PR TITLE
Xcode and testing upgrades

### LINKED ISSUE #82: Testing code - get it working again
Marking this as a bug. It really needs to happen.

### PR DESCRIPTION
**Pull request by bpbond**
This PR:

* Upgrades the Xcode project file to Xcode 11. More important, it changes things so that all the path settings (for header and library search) are defined using variables such as `BOOSTLIB`. These can be set by users in their Custom Path settings (Preferences -> Locations -> Custom Paths) to their particular machine values. This means _we won't be overwriting each other's settings_ as is currently happening.
* Starts to resurrect the C++ testing framework. There are two Xcode targets: `hector` (the default), and now `hector-tests`. The latter adds `${GTEST}` and `${GTESTLIB}` search paths, links against the gtest dynamic library, and uses `testing_main.cpp` instead of `main.cpp`. Currently I have only included the tseries test file; some work will be needed to bring each file up to date, as it's been a _long_ time since these tests were used.
* Updates the documentation on building using Xcode.

Still to do: add a GitHub Action that builds and runs the tests.

Closes #82
Closes #258

### CODE REVIEW

**Comment by bpbond** [comment]
@kdorheim I would suggest that we split this

>Still to do: add a GitHub Action that builds and runs the tests.

into a separate issue/PR, and merge this. Does that work for you?

**Comment by kdorheim** [comment]
> into a separate issue/PR, and merge this. Does that work for you?

Yes that works for me and makes a lot of sense especially since we want to revisit the github actions so to minimize their runtime cost

**Comment by bpbond** [comment]
Good point! OK, if things look good approve and merge, and then we can pull these changes into `stay-positive` etc.

File: vignettes/manual/BuildHector.Rmd [hunk 1]
```diff
@@ -135,37 +135,34 @@ Run Hector from the terminal with the command line.
...
+This means that developers and users can set these paths without overwriting each other.
 
-* *Architectures-Base SDK*: OS X 10.8 [or OS X version on machine] 
-* *Build Options-Compiler for C/C++/Objective-C*: Default compiler (Apple LLVM 5.1)
-* *Search Paths-Header Search Paths*: "/usr/local/include /usr/local/lib/boost_1_52_0/" [or other header directories, this is important for homebrew users]
-* *Search Paths-Library Search Paths*: "/usr/local/lib/" [or other library directories]
-* *LLVM-Language-C++-C++ Standard Library*: libc++ (LLVM C++ standard library with C++11 support)
-* *Linking-Other Linker Flags*: "-lgsl -lgslcblas -lm"
-* *User-Defined*: "GCC_MODEL_TUNING" defined as "G5"
+* You will need to set several custom paths (Preferences -> Locations -> Custom Paths): `BOOSTLIB`, `BOOSTROOT`, and `HECTORDIR`. These are the Boost runtime library, the Boost root under which the headers can be found in `boost/`, and the runtime working directory for the model (the repository root).
```
Comment (kdorheim) [inline vignettes/manual/BuildHector.Rmd:150]: I've followed these instructions but am running into issues with the boost finding some of the header files. 


Screen Shot 2021-06-24 at 11 16 18 AM


BOOSTROOT = `/usr/local/Cellar/boost/1.72.0_3/include` 

and that `BOOSTROOT/boost/math/tools/polynomial.hpp` does exist on my machine.... what am I doing wrong?? :(

**Review by kdorheim — CHANGES_REQUESTED** [review]

File: vignettes/manual/BuildHector.Rmd (hunk 1 above)
Comment (bpbond) [reply vignettes/manual/BuildHector.Rmd:150]: Verify that in $BOOSTROOT there is a `boost/` subdirectory—there is? And, inside of that, all the Boost hpp files? If not, you're pointing it to the wrong place.

If yes, you may need to revisit the build setting to force it to 'see' the proper location? Hmm.

File: inst/include/so2_component.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
+/* Hector -- A Simple Climate Model
```
[the remark this answers is not in the document]
Comment (bpbond) [reply inst/include/so2_component.cpp:1]: Uh...  🤦

File: vignettes/manual/BuildHector.Rmd (hunk 1 above)
Comment (kdorheim) [reply vignettes/manual/BuildHector.Rmd:150]: Okay so it turns out that you have to quit xcode and restart it in order for the custom paths to take effect 🤦‍♀️

File: vignettes/manual/BuildHector.Rmd [hunk 3]
```diff
@@ -135,37 +135,34 @@ Run Hector from the terminal with the command line.
...
+* If you want to build Hector's testing framework (the `hector-tests` target) add `GTEST` and `GTESTLIB` custom paths; these are analogous to their Boost counterparts.
+
```
Comment (kdorheim) [inline vignettes/manual/BuildHector.Rmd:152]: Dumb question but do I need to install gtest? Or some other software other than boost that has to be installed?

Comment (bpbond) [reply vignettes/manual/BuildHector.Rmd:152]: gtest is only needed if you want to build Hector's C++ testing framework.

Comment (kdorheim) [reply vignettes/manual/BuildHector.Rmd:152]: I do!

Comment (bpbond) [reply vignettes/manual/BuildHector.Rmd:152]: Oh, then, yes 😆

Comment (kdorheim) [reply vignettes/manual/BuildHector.Rmd:152]: YAY !!
```
[----------] Global test environment tear-down
[==========] 8 tests from 1 test suite ran. (0 ms total)
[  PASSED  ] 8 tests.
Program ended with exit code: 0
```

I would probably consider expanding that comment a bit then to make clear it is optional and that googletest would have to be installed (I was confused when googleing gtest lol)

If you want to build Hector's testing framework (this is optional and only relevant for C++ developers) add `GTEST` and `GTESTLIB` custom paths; these are analogous to their Boost counterparts. You will need to install googletest on your machine.

Comment (bpbond) [reply vignettes/manual/BuildHector.Rmd:152]: @kdorheim 👏 👏 👏 👏 👏 👏    nice!

Thanks—I will modify that.

File: inst/include/so2_component.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bpbond) [reply inst/include/so2_component.cpp:1]: I appear to have duplicated, not moved, it. Removing.

**Comment by bpbond** [comment]
@kdorheim All your comments have been addressed.

**Review by kdorheim — APPROVED** [review]

**Comment by kdorheim** [comment]
@bpbond go ahead and merge when you are ready, although is it going to be merged into stay positive or master?

**Comment by bpbond** [comment]
Yay, OK, thank you! I think it should go to `master` - no reason the 2.x Hector shouldn't take advantage of this.

**Pull request merged**